sql: separately measure the different phases of SQL query processing.

knz opened the pull request:
This patch samples the time at different stages of SQL execution.
The new latencies are reported at verbosity level 2 or above.

Example output, issuing a 4-way cross join from the CLI:
```
I170310 20:30:01.590757 302 sql/query_stats.go:71  [client=127.0.0.1:59248,user=root,n1] query stats: 625 rows, 0 retries, parse 124.76µs (2.9%), plan 149.59µs (3.5%), run 3945.72µs (91.4%), overhead 98.18µs (2.27%), batch age 4.322ms, session age 122.1948s
I170310 20:30:01.649721 302 sql/query_stats.go:71  [client=127.0.0.1:59248,user=root,n1] query stats: 1 rows, 0 retries, parse 5.98µs (6.5%), plan 52.99µs (57.5%), run 6.91µs (7.5%), overhead 26.28µs (28.51%), batch age 0.093ms, session age 122.2538s
```

Fixes #14019.
First step towards #13968.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14080)
<!-- Reviewable:end -->

knz committed:
sql: separately measure the different phases of SQL query processing.

This patch samples the time at different stages of SQL execution.
The new latencies are reported at verbosity level 2 or above.

knz commented:
cc @petermattis

knz commented:
FWIW here are the stats on a single-row lookup on a KV table using the PK:
```
query stats: 1 rows, 0 retries, parse 68.70µs (8.1%), plan 173.41µs (20.6%), run 509.05µs (60.3%), overhead 92.68µs (10.98%)
```

(we got work to do!)

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I didn't look at all of the places you're recording times. 

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/query_stats.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-KetwYLqXbTc6tqWv7xj:-KetwYLqXbTc6tqWv7xk:b5z54p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d497f1dd6bc9036099c79793f40e60cab12d73/pkg/sql/query_stats.go#L66)):*
> ```Go
> 	// Compute the run latency. This is always recorded in the
> 	// server metrics.
> 	runLat := phaseTimes[sessionEndExecStmt].Sub(phaseTimes[sessionStartExecStmt]).Nanoseconds()
> ```

Rather than convert to nanoseconds, I'd leave this as `time.Duration` (which is itself in nanoseconds). FWIW, when I'm formatting durations I'll do something like `runLat.Seconds() * 1000` to get milliseconds or `runLat.Seconds()*1e6` to get microseconds. Note that `Seconds()` returns a `float64`. That might not work well given the percentages you're computing.

---

*[pkg/sql/query_stats.go, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-KetwNsTMObAchXK46TX:-KetwNsTMObAchXK46TY:b-nkn6iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d497f1dd6bc9036099c79793f40e60cab12d73/pkg/sql/query_stats.go#L80)):*
> ```Go
> 	// TODO(knz) #13968.
> 	if log.V(2) {
> 		runLatf := float64(runLat)
> ```

The `f` suffix is confusing since my brain reads it is `format`, not `float`. And you don't seem to use it consistently (i.e. `batchAge`).

---

*[pkg/sql/query_stats.go, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-KetwCuGerz5VR1CHJjH:-KetwCuGerz5VR1CHJjI:bvjgfnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d497f1dd6bc9036099c79793f40e60cab12d73/pkg/sql/query_stats.go#L110)):*
> ```Go
> 			"plan %.2fµs (%.1f%%), "+
> 			"run %.2fµs (%.1f%%), "+
> 			"overhead %.2fµs (%.2f%%), "+
> ```

Why does this percentage get 2 decimal places while the others get 1?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-:-KetxNvptTlmEY1BHZ9k:b-wo2foe)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/query_stats.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-KetwYLqXbTc6tqWv7xj:-KeuDDYm-xMKRx6Nu9vN:b-tl9jwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d497f1dd6bc9036099c79793f40e60cab12d73/pkg/sql/query_stats.go#L66)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Rather than convert to nanoseconds, I'd leave this as `time.Duration` (which is itself in nanoseconds). FWIW, when I'm formatting durations I'll do something like `runLat.Seconds() * 1000` to get milliseconds or `runLat.Seconds()*1e6` to get microseconds. Note that `Seconds()` returns a `float64`. That might not work well given the percentages you're computing.
</blockquote></details>

Done.
Percentages are ok for now.

---

*[pkg/sql/query_stats.go, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-KetwNsTMObAchXK46TX:-KeuDFm8Gff3EIwgUUSG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d497f1dd6bc9036099c79793f40e60cab12d73/pkg/sql/query_stats.go#L80)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `f` suffix is confusing since my brain reads it is `format`, not `float`. And you don't seem to use it consistently (i.e. `batchAge`).
</blockquote></details>

Done.

---

*[pkg/sql/query_stats.go, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-KetwCuGerz5VR1CHJjH:-KeuDGImOnKig3v3-ilF:b7e94k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d497f1dd6bc9036099c79793f40e60cab12d73/pkg/sql/query_stats.go#L110)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why does this percentage get 2 decimal places while the others get 1?
</blockquote></details>

Oversight. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1410 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Oa0P_9ykzkC_bVMJ:-Kf7Oa0P_9ykzkC_bVMK:b-6npo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cea587167caa9a6a260b279f5f7baca7e14f292/pkg/sql/executor.go#L1410)):*
> ```Go
> 		return result, err
> 	}
> 	if useDistSQL && !isAutomaticRetry {
> ```

do you know what `!isAutomaticRetry` is about? Why don't we use DistSQL on automatic retries?

---

*[pkg/sql/executor.go, line 1426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7MRirqW-Yjm8l3t4E:-Kf7MRirqW-Yjm8l3t4F:bhz7t8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/executor.go#L1426)):*
> session.phaseTimes[:]
I didn't know what this meant. Please comment the arg saying that we're turning the array to a slice.
Or try to pass it by pointer-to-fixed-len-array. And explain in the method comment that the param is const but we want to avoid the copy. I think that'd be better as we'd be sticking to a single type for this array, and we could give that type an alias.

---

*[pkg/sql/query_stats.go, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7NoreYd6cjXMN6m61:-Kf7NorgX2Mfp-6vWelV:b-1fy76) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L1)):*
> ```Go
> // Copyright 2015 The Cockroach Authors.
> ```

how about naming this file `executor_query_stats.go` (or `executor_statement_metrics`, ideally the name would be related to the method's name), analogous to `replica_range_lease.go` (signifying that the code in it is part of the`Executor` type)

---

*[pkg/sql/query_stats.go, line 31 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7JmSihEZeBG_OVu56:-Kf7JmSihEZeBG_OVu57:b-zvjf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L31)):*
> ```Go
> 
> const (
> 	sessionInit sessionPhase = iota
> ```

perhaps comments on the first two hinting about what's encompassed in the "init" would be helpful

---

*[pkg/sql/query_stats.go, line 44 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7KljtoChE3w8fcZ6M:-Kf7KljtoChE3w8fcZ6N:b-jhb6xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L44)):*
> ```Go
> )
> 
> func (e *Executor) recordStatementMetrics(
> ```

this method deserves a comment

---

*[pkg/sql/query_stats.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Okod1dv6buUKrsTb:-Kf7Okoek3NYoF7x3xHG:bta0zj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L48)):*
> ```Go
> 	stmt parser.Statement,
> 	phaseTimes []time.Time,
> 	distributedQuery bool,
> ```

I think `distributedQuery` (which perhaps should be more explicitly named - `distSQLUsed`) and `retryCount` deserve comments. Does the `retry` only refer to automatic retries?

---

*[pkg/sql/query_stats.go, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7KAsupfi2uAhmeNs9:-Kf7KAsvrEVreQPNeoon:bewg45j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L54)):*
> ```Go
> ) {
> 
> 	// SQL execution is separated in 3+ phases:
> ```

seems to me that this comment should be lifted as either a method comment or a file-level comment.

---

*[pkg/sql/session.go, line 115 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7NGGlP0RlnhZWKda8:-Kf7NGGmXk_wBXF8FZSF:b-1qhl6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/session.go#L115)):*
> ```Go
> 	// statistics to measure the time spent in each
> 	// phase of SQL execution.
> 	phaseTimes [sessionNumPhases]time.Time
> ```

I think this guy deserves a type alias (if we no longer covert it to a different type (slice) when we pass it around).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-:-Kf7P-KzdFL5OHLRciH3:bb76hxo)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/executor.go, line 1410 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Oa0P_9ykzkC_bVMJ:-Kf7yhJ8D-vE5ZsDyfyE:blhp485) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cea587167caa9a6a260b279f5f7baca7e14f292/pkg/sql/executor.go#L1410)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do you know what `!isAutomaticRetry` is about? Why don't we use DistSQL on automatic retries?
</blockquote></details>

No I do not know. But if we want to change this it would need to be a different PR :stuck_out_tongue: 

---

*[pkg/sql/executor.go, line 1426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7MRirqW-Yjm8l3t4E:-Kf80A-8a8sfRdACJUit:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/executor.go#L1426)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> session.phaseTimes[:]
I didn't know what this meant. Please comment the arg saying that we're turning the array to a slice.
Or try to pass it by pointer-to-fixed-len-array. And explain in the method comment that the param is const but we want to avoid the copy. I think that'd be better as we'd be sticking to a single type for this array, and we could give that type an alias.
</blockquote></details>

Done.

---

*[pkg/sql/session.go, line 115 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7NGGlP0RlnhZWKda8:-Kf8-uE4Aj75OOPRDn5e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/session.go#L115)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this guy deserves a type alias (if we no longer covert it to a different type (slice) when we pass it around).
</blockquote></details>

Done.

---

*[pkg/sql/query_stats.go, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7NoreYd6cjXMN6m61:-Kf7zkKR4_GUwucMun6M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L1)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how about naming this file `executor_query_stats.go` (or `executor_statement_metrics`, ideally the name would be related to the method's name), analogous to `replica_range_lease.go` (signifying that the code in it is part of the`Executor` type)
</blockquote></details>

Done.

---

*[pkg/sql/query_stats.go, line 31 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7JmSihEZeBG_OVu56:-Kf7zwYqsCd2CFuya0Sp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L31)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

perhaps comments on the first two hinting about what's encompassed in the "init" would be helpful
</blockquote></details>

Done.

---

*[pkg/sql/query_stats.go, line 44 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7KljtoChE3w8fcZ6M:-Kf8-AL8gILqvHx8Jxal:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L44)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this method deserves a comment
</blockquote></details>

Done.

---

*[pkg/sql/query_stats.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Okod1dv6buUKrsTb:-Kf8-Trw2ZQOlWd-zg3Q:brodup0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think `distributedQuery` (which perhaps should be more explicitly named - `distSQLUsed`) and `retryCount` deserve comments. Does the `retry` only refer to automatic retries?
</blockquote></details>

I do not know personally what these things really mean, so I am unable to document them myself. Also I do not think the reader of this code should really care -- the role of this method is merely to save up and report on the variables used in the rest of the executor code; the interpretation of this information is fully delegated to the person who will use these stats later.

---

*[pkg/sql/query_stats.go, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7KAsupfi2uAhmeNs9:-Kf8-i0TNEG3Mv5Q__zR:bohqzsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/query_stats.go#L54)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

seems to me that this comment should be lifted as either a method comment or a file-level comment.
</blockquote></details>

Agreed, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I think you haven't pushed your changes

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1410 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Oa0P_9ykzkC_bVMJ:-Kf86FHnlhLN30mR7TXP:b-1253to) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cea587167caa9a6a260b279f5f7baca7e14f292/pkg/sql/executor.go#L1410)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No I do not know. But if we want to change this it would need to be a different PR :stuck_out_tongue: 
</blockquote></details>

https://github.com/cockroachdb/cockroach/pull/14116

---

*[pkg/sql/executor_statement_metrics.go, line 1 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7NoreYd6cjXMN6m61:-Kf80YTOPEYjbVJlFU9a:becxvly) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/executor_statement_metrics.go#L1)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

woohoo Reviewable undestands renames now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-:-Kf86CYxjictOqNPhEaa:b-te5tob)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1410 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Oa0P_9ykzkC_bVMJ:-Kf8B-rCuB0XWO2vhkRZ:bepw6vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cea587167caa9a6a260b279f5f7baca7e14f292/pkg/sql/executor.go#L1410)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/pull/14116
</blockquote></details>

ok can you merge that then I'll pick it up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080)*
<!-- Sent from Reviewable.io -->

knz commented:
Indeed I had forgotten the commit. PTAL.

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-:-Kf8Gfmpmg1x9lOqhEgu:b-1jqsld)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 643 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8eBgrkEc22wyxiAYT:-Kf8eBgrkEc22wyxiAYU:bhg9e3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor.go#L643)):*
> ```Go
> 
> 		// Track if we are retrying this query, so that we do not double count.
> 		retryCount := 0
> ```

I think `automaticRetryCount` would be a better name here in and in the new method param - the distinction between automatic and user retries relevant to the new metrics.

---

*[pkg/sql/executor.go, line 1042 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8cSk9ObrpcbAwvvSg:-Kf8cSkAbmi2TRIqOi-F:b-l9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor.go#L1042)):*
> ```Go
> // stmtTimestamp: Used as the statement_timestamp().
> // retryCount: increases with each retry; 0 for the first attempt
> // count in metrics.
> ```

second line of the comment is now stray. And I think `automaticRetryCount` is better.

---

*[pkg/sql/executor_statement_metrics.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Okod1dv6buUKrsTb:-Kf81CpcvvgD5MsoVAXH:b-arj9yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/executor_statement_metrics.go#L48)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I do not know personally what these things really mean, so I am unable to document them myself. Also I do not think the reader of this code should really care -- the role of this method is merely to save up and report on the variables used in the rest of the executor code; the interpretation of this information is fully delegated to the person who will use these stats later.
</blockquote></details>

:) I call "not reasonable" :)

You're creating this method, taking variables out of their old context, you got to be able to describe the arguments. Only in broader refactoring situations can you play the "not my code" card.
Also, you've introduced `retryCount` in this PR (it used to be a narrower bool). We should document that we don't count automatic retries in whatever metric.

Specifically, I asked for comments because these arguments are surprising for a method that "records metrics". Perhaps I wouldn't have said anything if the method name also had to do with the logging - which is the reason why `retryCount` was made an int from a bool and passed here. Maybe `recordStatementStats`, but now I'm nitpicking.

---

*[pkg/sql/executor_statement_metrics.go, line 72 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8dDtWWWJNCaVBPXx7:-Kf8dDtXiT0aPIs904iB:b-e5yp5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor_statement_metrics.go#L72)):*
> ```Go
> 	stmt parser.Statement,
> 	phaseTimes *phaseTimes,
> 	useDistSQL bool,
> ```

nit: I think `distSQLUsed` (non-imperative form) would be better, since it reflects the history not the future. You can comment that distsql vs non-distsql metrics are incremented based on this. This comment is not particularly important though. The `retryCount` arg is more surprising.

---

*[pkg/sql/executor_statement_metrics.go, line 82 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8bX51AFepR-hiinzs:-Kf8bX51AFepR-hiinzt:b-nefwi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor_statement_metrics.go#L82)):*
> ```Go
> 	runLatRaw := phaseTimes[sessionEndExecStmt].Sub(phaseTimes[sessionStartExecStmt])
> 
> 	if useDistSQL {
> ```

I think here you want to only count if the retryCount is 0 (to not double count). That's what the old code was doing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 643 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8eBgrkEc22wyxiAYT:-Kf8gLySZ8qFHkF6PZiM:b-el989i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor.go#L643)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think `automaticRetryCount` would be a better name here in and in the new method param - the distinction between automatic and user retries relevant to the new metrics.
</blockquote></details>

I agree. Done.

---

*[pkg/sql/executor.go, line 1042 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8cSk9ObrpcbAwvvSg:-Kf8gYFesp1hlRa-agsz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor.go#L1042)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

second line of the comment is now stray. And I think `automaticRetryCount` is better.
</blockquote></details>

Done.

---

*[pkg/sql/executor_statement_metrics.go, line 48 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf7Okod1dv6buUKrsTb:-Kf8hAAmUUAI0lnXZPU5:b-jugu30) ([raw file](https://github.com/cockroachdb/cockroach/blob/a60b178cb1fd6ccc98076186eccc65926fa72320/pkg/sql/executor_statement_metrics.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

:) I call "not reasonable" :)

You're creating this method, taking variables out of their old context, you got to be able to describe the arguments. Only in broader refactoring situations can you play the "not my code" card.
Also, you've introduced `retryCount` in this PR (it used to be a narrower bool). We should document that we don't count automatic retries in whatever metric.

Specifically, I asked for comments because these arguments are surprising for a method that "records metrics". Perhaps I wouldn't have said anything if the method name also had to do with the logging - which is the reason why `retryCount` was made an int from a bool and passed here. Maybe `recordStatementStats`, but now I'm nitpicking.
</blockquote></details>

The rename is warranted. Changed to `recordStatementSummary`.
I have added some comments too.

---

*[pkg/sql/executor_statement_metrics.go, line 72 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8dDtWWWJNCaVBPXx7:-Kf8hOlgJhH2VD5NVDQg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor_statement_metrics.go#L72)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I think `distSQLUsed` (non-imperative form) would be better, since it reflects the history not the future. You can comment that distsql vs non-distsql metrics are incremented based on this. This comment is not particularly important though. The `retryCount` arg is more surprising.
</blockquote></details>

Done.

---

*[pkg/sql/executor_statement_metrics.go, line 82 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-Kf8bX51AFepR-hiinzs:-Kf8gooGYL-gKbBBVcWl:b-u77ywr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dfb656641fefefbf3572ce8179a8d0dd06fe798/pkg/sql/executor_statement_metrics.go#L82)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think here you want to only count if the retryCount is 0 (to not double count). That's what the old code was doing.
</blockquote></details>

Indeed! Thanks for spotting this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks

---

Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14080#-:-Kf8lCvVCPUpIMfLqIbK:b-v6ynew)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR!
